docs/design/core/metrics/prototype/MetricPublisher.java

```diff
@@ -20,39 +20,72 @@
+ * multithreading, it's possible that the publisher is shared concurrently by
+ * multiple threads, and necessitates that all implementations are threadsafe.
+ * <p>
+ * <b>Example:</b>
+ * At {@code t0}:
+ * {@code
+ *     metricEventsBuilder.putMetricEvent(Events.MARSHALLING_START, Instant.now());
+ * }
  * <p>
- *     In certain situations (high throttling errors, metrics are reported faster than publishing etc), storing all the metrics
- *     might take up lot of memory and can crash the application. In these cases, it is recommended to have a max limit on
- *     number of metrics stored or memory used for metrics and drop the metrics when the limit is breached.
- * </p>
+ * At {@code t1}, after mashalling is complete:
+ * {@code
+ *     metricEventsBuilder.putMetricEvent(Events.MARSHALLING_END, Instant.now());
+ * }
+ * <p>
+ * At {@code t2} after the SDK operation is complete:
+ * <p>
+ * {@code
+ *     metricPublisher.consume(metricEventsBuilder.build());
+ * }
+ * At some later {@code tN}, the publisher can then choose to aggregate all of
+ * the {@code metricEvents} it has received and publish them.
+ *
+ *
+ * Implementations must be threadsafe.
  */
 @SdkPublicApi
+@ThreadSafe
 public interface MetricPublisher extends AutoCloseable {
     /**
-     * Publish all metrics stored in the publisher. If all available metrics cannot be published in a single call,
-     * multiple calls will be made to publish the metrics.
-     *
-     * It is recommended to publish the metrics in a non-blocking way. As it is common to publish metrics to an external
-     * source which involves network calls, the method is intended to be implemented in a non-blocking way and thus
-     * returns a {@link CompletableFuture}.
-     *
-     * Depending on the implementation, the metrics are published to the external source periodically like:
-     * a) after a certain time period
-     * b) after n metrics are registered
-     * c) after the buffer is full
+     * Notify the publisher of new metric data. After this call returns, the
+     * caller can safely discard the given {@code metricEvents} instance if it
+     * no longer needs it. Implementations are strongly encouraged to complete
+     * the aggregation and publishing of metrics in an asynchronous manner to
+     * avoid blocking the calling thread.
+     * <p>
+     * With the exception of a {@code null} {@code metricEvents}, all
+     * invocations of this method must return normally. The only legal way to
+     * report an error is by completing the returned future exceptionally. This
+     * is to ensure that callers of the publisher can safely assume that even
+     * in situations where an error happens during publishing that it will not
+     * interrupt the calling thread.
+     * <p>
+     * The future is completed when the metrics calculated or otherwise derived
+     * from the given {@code metricEvents} have been published.
```

Comment:
That sounds like an ambitious contract, are we going to keep track of each completable future that acts as a component for each aggregated metric and then make sure we keep track of it all when we publish? What if the publisher is complex and publishes to other publishers or something like that? Do we need to return anything at all? What is the caller going to do with the information?

Reply:
I don't really think it's necessary to return anything. One use case I can think if is preventing JVM exit by waiting on the futures returned but we also enforce flushing before returning in the `close()` method, so I think it's actually fine for this to be `void`.